Functor Slides

JorgeCastilloPrz opened the pull request:
Fixes #32 

Here you have a first version of the `Functor` slides. Please suggest any further changes or anything you still find missing.

JorgeCastilloPrz commented:
@raulraja I've not appended anything about bindings since map is not very relevant to sequential computations but flatMap. Let me know if you miss anything else, pls.

raulraja commented on Functor.md:
````diff
@@ -0,0 +1,150 @@
+    //...
+}
+```
+* The `F` on the declaration stands for a **type constructor**. That's why `Functor`is considered **polymorphic**.
````
> That's why `Functor`is considered **polymorphic**.
I would say here instead:
> Functor is parametric over a type constructor. This allows us to build abstract functions over the behaviors that Functor defines and forgetting about the concrete types that `F` may refer to. We call this ad-hoc polymorphism, the ability to write polymorphic programs that can be defined in generic terms.

raulraja commented on Functor.md:
```diff
@@ -0,0 +1,150 @@
+* ...and many more.
+
+# Slide 7
+You can request the functor instance statically from any type constructor supporting it:
```
The alternative to request the function statically and the most important perhaps to showcase in this slides is the ability to write polymorphic programs that are abstract and then made concrete:
```kotlin
fun <F> Functor<F>.addOne(fa: Kind<F, Int>): Kind<F, Int> =
  fa.map { it + 1 }

Option.functor().addOne(Option(1)).fix() // Option(2)
Try.functor().addOne(Try { 1 }).fix() // Success(2)
```

raulraja commented on Functor.md:
```diff
@@ -0,0 +1,150 @@
+
+# Slide 10
+`Option` just contains a value when its type is `Some<A>`. So we can `map` over its content just for that case.
+Because of that, we say that `Option` is a *"biased"* to the case when it contains an actual value.
```
... `Option` is biased toward the `Some` case...

raulraja commented on Functor.md:
````diff
@@ -0,0 +1,150 @@
+none.map { it + 1 }
+// None
+```
+As you can see, the error gets short-circuited when it's a `None`.
````
s/error/computation

raulraja commented on Functor.md:
```diff
@@ -0,0 +1,150 @@
+As you can see, the error gets short-circuited when it's a `None`.
+
+# Slide 11
+Same thing happens for other data types like `Try<A>`. `Try` wraps a potentially throwing computation, and it is defined like:
```
Instead of `wraps a potentially throwing computation` I'd say `models safe access to API's that may throw exceptions`

raulraja commented on Functor.md:
````diff
@@ -0,0 +1,150 @@
+# Slide 12
+`Try` is biased to its `Success` implementation, so map just works over that one. Otherwise it short-circuits the error:
+```
+val failingOp = Try { failingOperation() }
````
I think all these examples should be done in polymorphic functions and not on the actual data types then made concrete on the data types to show the real power of type classes which is where Functor and Type classes really make sense and that is to provide the ability to write polymorphic programs.

raulraja commented on Functor.md:
````diff
@@ -0,0 +1,153 @@
+    //...
+}
+```
+* Functor is parametric over a type constructor `F`. This allows us to build abstract functions over the behaviors that Functor defines and forgetting about the concrete types that F may refer to.
````
👏

raulraja commented on Functor.md:
```diff
@@ -0,0 +1,153 @@
+
+# Slide 4
+`Functor` abstracts the hability to **map** over the computational context of a type constructor `F`.
+In other words, it provides a mapping function for the `F` type:
```
👏

JorgeCastilloPrz commented:
Alright, working on it.

JorgeCastilloPrz commented:
* Added slides showcasing the laws that must be satisfied by any `Functor` instance to be considered a `Functor`.
* Added an example of how you can test your own functor instances by importing the `FunctorLaws` from `arrow-test` into your testing environment.
* Reformatted existing slides a bit.

Pls @raulraja, one more time? :)

Also, I'm wondering if .MD file should go in the root dir or not. I've noticed it's not very similar to the ones we have there atm. Is the format ok on this one? i.e: I didn't provide slide titles but just the content following the style I found in #15 , which I don't know if it's correct or not. I assume each slide should have a title about what's it talking about, instead of "Slide X"

Also I'm not sure if this would be just slides or also used as the script, or how you use to proceed on this. Should I go for a different file for the script after this one?

raulraja commented:
The slides should be the same as in Option or OptionT since this will render on Deckset for Mac for the video

JorgeCastilloPrz commented:
I've formatted the slides for deckset now and removed intro and final sections since those where already recorded in SF.

JorgeCastilloPrz commented:
One final 👍 before merging @raulraja ? I've tested it on Deckset to ensure it looks good.

benjymontoya commented:
@MaureenElsberry everything is ready for the lang review. Thanks!

JorgeCastilloPrz commented:
Working on the script now on this PR.

JorgeCastilloPrz commented:
Hi @benjymontoya. I've added the script after each one of the slides now. Could you take a look one more time @raulraja and @MaureenElsberry for lang corrections please?

I'm trying to move faster towards publishing this video in the following weeks.

Thanks guys.

raulraja commented on Functor.md:
````diff
@@ -0,0 +1,340 @@
+In other words, it provides a mapping function for the `F` type:
+
+```kotlin
+fun F<A>.map(f: (A) -> B): F<B>
````
I think these should use `Kind<F, A>` instead of `F<A>` since `F<A>` is not valid kotlin syntax and will only confuse viewers.

JorgeCastilloPrz replied:
Fixed.

raulraja commented on Functor.md:
````diff
@@ -0,0 +1,340 @@
+    fa.map(f).map(g) = fa.map(f andThen g)
+    ```
+
+^ Functor satisfies two laws: Identity and Composition.
````
These descriptions are excellent! Kudos :clap:

JorgeCastilloPrz commented:
Thanks for your help @MaureenElsberry :). If I have green light I'll start recording this week.

JorgeCastilloPrz commented:
Audio recording of the script is complete @benjymontoya @raulraja. I'm sending it to both of you guys by Slack (since it's high quality audio and the file is around 40 mb and github does not accept files higher than 10 mb.

JorgeCastilloPrz commented:
Merging this since it was already released.
